PICARD-1926: Use for per-thread config instances

phw opened the pull request:
<!--
    Hello! Thanks for submitting a pull request to MusicBrainz Picard. We
    appreciate your time and interest in helping our project!

    Use this template to help us review your change. Not everything is required,
    depending on your change. Keep or delete what is relevant for your change.
    Remember that it helps us review if you give more helpful info for us to
    understand your change.

    Ensure that you've read through and followed the Contributing Guidelines, in
    [CONTRIBUTING.md](https://github.com/metabrainz/picard/blob/master/CONTRIBUTING.md).
-->

# Summary

<!--
    Update the checkbox with an [x] for the type of contribution you are making.
-->

* This is a…
    * [x] Bug fix
    * [ ] Feature addition
    * [ ] Refactoring
    * [ ] Minor / simple change (like a typo)
    * [ ] Other
* **Describe this change in 1-2 sentences**:

# Problem

<!--
    Anything that helps us understand why you are making this change goes here.
    What problem are you trying to fix? What does this change address?
-->

* JIRA ticket (_optional_): PICARD-1926, PICARD-1689, PICARD-2013, ~PICARD-1590~
<!--
    Please make sure you prefix your pull request title with 'PICARD-XXX' in order
    for our ticket tracker to link your pull request to the relevant ticket.
-->

QSettings is not thread safe, but instead each thread must use its own instance of QSettings. Otherwise this leads to deadlocks when reading settings shortly after writing settings in different threads.


# Solution
Prepare per-thread config instances by adding config.get_config() and using it to fix deadlocks experienced in the metadatabox.

<!--
    The details of your change. Talk about technical details, considerations, or
    other interesting points. If you have a lot to say, be more detailed in this
    section.
-->


# Action

<!--
    Other than merging your change, do you want / need us to do anything else
    with your change? This could include reviewing a specific part of your PR.
-->

phw commented:
@zas This is WIP and tests need fixing, but I would appreciate your early feedback on this. This would need to get rolled out over the entire code base.

The main threads config instance stays available as `config.config` (and `config.settings` or `config.persist`) for backward compatibility, especially plugins. For the most parts we don't run into issues as we only read from config. The issue really only arises if reading happens shortly after writing from different threads.

We tried to patch around this a lot: We had a fix for similar PICARD-1590 which we likely can remove then. We have the locks in `ConfigSection` which try to make this thread  safe, but fail because QSettings does it's work delayed and asynchronous. Probably these locks also can be removed if we make use of `get_config`.

zas commented on picard/config.py:
```diff
@@ -341,12 +336,51 @@ class ListOption(Option):
+    """Returns a config object for the current thread.
+
+    Config objects for threads are
+    """
```
Overall the patch looks good, but I think we should totally get rid of global `config` (also `setting` and `persist`), because keeping those along `get_config()` is rather weird and prolly prone to errors.
It would be great if `get_config()` could work in any context (threads or not).

phw replied:
get_config works in any context, you totally should use it also in the main thread. The idea to keep the main thread's config still available as config.config is really for backward compatibility. If we remove it we break a lot of plugins.

zas replied:
Ah ok, it makes sense.

phw replied:
And it also allows me to convert the code base step by step while still having a working system. I hope that makes it easier to update some parts of the code and test it. For starters I'll focus on those parts where we actually run threaded.

zas commented on picard/config.py:
```diff
@@ -341,12 +319,51 @@ class ListOption(Option):
+        for thread_id in list(_thread_configs.keys()):
+            if thread_id not in all_threads:
+                log.debug('Purging config instance for thread %s.', thread_id)
+                del _thread_configs[thread_id]
```
This can perhaps be rewriiten so it uses only sets:

```python
        all_threads = set([thread.ident for thread in threading.enumerate()])
        threads_config = set(_thread_configs)
        for thread_id in threads_config.difference(all_threads):
            log.debug('Purging config instance for thread %s.', thread_id)
            del _thread_configs[thread_id]
```

phw replied:
Not sure, I don't really think that it is easier to read.

zas replied:
Yup, not easier to read for sure, but likely faster. You can still rewrite `list(_thread_configs.keys())` to `list(_thread_configs))`.

zas commented on picard/config.py:
```diff
@@ -341,12 +319,51 @@ class ListOption(Option):
+    return thread_config
+
+
+def purge_config_instances():
```
Since this method is only called from `setup_config()` it should perhaps be renamed `_purge_config_instances()`

zas commented on picard/config.py:
```diff
@@ -341,12 +319,51 @@ class ListOption(Option):
 setting = None
 persist = None
 
+PURGE_INTERVAL = 60000
```
Since `PURGE_INTERVAL` is a number of milliseconds (according to https://doc.qt.io/qt-5/qtimer.html) it would be cleared to include unit in the name: `PURGE_INTERVAL_MILLISECONDS`.

zas commented on picard/config.py:
```diff
@@ -341,12 +319,51 @@ class ListOption(Option):
+    thread_config = _thread_configs.get(thread_id)
+    if not thread_config:
+        _thread_config_lock.acquire()
+        try:
```
Is there a specific reason to use try/finally here? It may hide exceptions.

Is it possible to just use:

```python
with _thread_config_lock:
    do_stuff
```

and if exceptions are expected:

```python
with _thread_config_lock:
    try:
        do_stuff
    except Exception as e:
        log_this
```

phw replied:
The locks from threading do not implement a context manager, so you cannot use `with`. But

```python
_thread_config_lock.acquire()
try:
    do_stuff
finally:
    _thread_config_lock.release()
```

is functional equivalent to

```python
with _thread_config_lock:
    do_stuff
```

It doesn't hide exceptions, as it doesn't catch any. But it ensures the lock is released. It shouldn't matter here. No exceptions are expected here, and should one arise the main thread would exit anyway, so no real need to release the lock. But I think it is good practice to follow this pattern when dealing with locks.

zas replied:
Damn, I shouldn't review stuff without enough coffee in my body...

phw commented:
I addressed the comments. Currently the state of this should be fine to fix the deadlock issues we had, as far as I can tell all uses of the config instance inside threads now use a thread specific `Config` instance. Things to do:

1. Tests are still failing sometimes, and I don't yet know why. This needs to be fixed before merging.

2. For consistency we should also convert other uses and getting rid of direct access to `config.config` in Picard's codebase (but as discussed above keep it for compatibility with plugins). We could do the rest of the conversion in a separate PR in order to not blow this one up too much.

3. I will review existing plugins. If a plugin accesses config inside a thread I'll refactor it in a way that keeps backward compatibility. Either move the config access outside of the thread or try to import `get_config` with a fallback that will just do
    ```python
    def get_config():
       return config.config
    ```

phw commented:
One thing I clearly noticed during my tests is that instantiating a `QSettings` instance is really slow. When originally adapting the tests to the new config handling I actually used a real `QSettings` instance (with a temporary file) per test. This made the tests run 3 times slower. So we really should avoid creating the `Config` object too often. But this patch ensures this by having only one instance per thread and keeping it around at least as long as the thread is active.

One thing is that I measured a slow down of file loading performance. Instead of 4k files getting loaded in ~13.6 seconds it now took ~15.4 seconds. I am not sure yet on the cause. It is not the instantiation of `Config` objects. This only happens 8 times (`QThreadPool` provides 8 threads on my system due to available CPU cores). So it is probably the repeated calls to `get_config`.

zas commented on picard/config.py:
```diff
@@ -109,32 +105,27 @@ def raw_value(self, name, qtype=None):
+            except Exception as why:
+                log.error('Cannot read %s value: %s', self.key(name), why, exc_info=True)
+                value = default
+            self.unlock()
```
Use `finally` here to ensure `unlock()` is called whatever happens?

zas commented on picard/ui/metadatabox.py:
```diff
@@ -132,21 +136,24 @@ def __tag_ne(self, tag, orig, new):
             self.new.add(tag, new_values)
 
+        if not top_tags:
+            top_tags = []
```
Instead of a list, since it is only used below as `tag in top_tags`, and never modified, we could use (and pass a set).

zas commented on picard/ui/metadatabox.py:
```diff
@@ -566,12 +576,14 @@ def _update_tags(self, new_selection=True, drop_album_caches=False):
         tag_diff.objects = len(files)
 
         clear_existing_tags = config.setting["clear_existing_tags"]
+        top_tags = config.setting['metadatabox_top_tags']
```
`top_tags` -> set?

zas reviewed (commented):
Overall, I like this move

zas commented:
Weird, tests are segfaulting with 384f2cf

phw commented:
> Weird, tests are segfaulting with [384f2cf](https://github.com/metabrainz/picard/commit/384f2cfb0ecc25731c8c2bfded2243d30c41e51f)

The reason is that it calls the parent constructor in __init__. Without this it doesn't crash. Actually before https://github.com/metabrainz/picard/commit/ebd98c9758d074804c2dd83059827f31fbdb13fd#diff-74efa3849396e3135d3796b5965e96c02fcc9dc1036c4b28bba71506193303d6L109 we had only an empty `__init__` method.

I am not sure what consequences it has to omit the QSettings constructor, but it seems to work.

Sophist-UK commented:
{Heresy alert}

I know this is a bit off the wall (only a bit?) but this seems to be a way complicated solution to a thread issue with Qt/PyQt.

Can we not find a native thread aware solution in Python that is way easier?

phw commented:
It just uses QSettings as it is intended to be used, I don't see it as overly complicated. Do you have any suggestion of another solution that is simpler and both syncs between threads and processes?

zas commented on picard/config.py:
```diff
@@ -341,12 +322,51 @@ class ListOption(Option):
+            del _thread_configs[thread_id]
+    finally:
+        _thread_config_lock.release()
+        QtCore.QTimer.singleShot(PURGE_INTERVAL_MILLISECONDS, _purge_config_instances)
```
What about something like:

```python
def _purge_configs_timer_start(purge_interval_milliseconds=60000):
    def _purge_config_instances():
        """Removes cached config instances for no longer active threads."""
        _thread_config_lock.acquire()
        try:
            all_threads = set([thread.ident for thread in threading.enumerate()])
            threads_config = set(_thread_configs)
            for thread_id in threads_config.difference(all_threads):
                log.debug('Purging config instance for thread %s.', thread_id)
                del _thread_configs[thread_id]
        finally:
            _thread_config_lock.release()
            _purge_configs_timer_start(purge_interval_milliseconds)

    QtCore.QTimer.singleShot(purge_interval_milliseconds, _purge_config_instances)
```

phw replied:
Instead of defining a new `_purge_config_instances` per call I'd rather put this on module level, at which point `_purge_configs_timer_start` is just a call to `QTimer.singleShot`. Maybe if you want to have the interval encapsulated something like:

```
def _init_purge_config_timer(purge_interval_milliseconds=60000):
    def purge_config_instances():
        """Removes cached config instances for no longer active threads."""
        _thread_config_lock.acquire()
        try:
            all_threads = set([thread.ident for thread in threading.enumerate()])
            threads_config = set(_thread_configs)
            for thread_id in threads_config.difference(all_threads):
                log.debug('Purging config instance for thread %s.', thread_id)
                del _thread_configs[thread_id]
        finally:
            _thread_config_lock.release()
            start_purge_config_timer()

    def start_purge_config_timer():
        QtCore.QTimer.singleShot(purge_interval_milliseconds, purge_config_instances)

    start_purge_config_timer()
```

phw replied:
What do you think about 8053e59 ?

phw commented:
Tested this again with the docker-image, where I could reliable reproduce the freezes. They are completely gone with this fix.

RobertKrawitz commented:
I think this fix is still incomplete.  To reproduce: select all items in an album with multiple tracks along with the album itself, do Add tag.  When I then click Save, it locks up in the same way.

(I wish Picard let you save the current state of processing rather than just save the files; I'm going to have to redo all my work as a result.)

phw commented:
@RobertKrawitz Are you testing the current master branch?

RobertKrawitz commented:
Sorry, missed the notification email on this -- I filed a jira ticket a few days ago that was on the latest (at the time) master branch, post-this merge

phw commented:
@RobertKrawitz All good, we have been discussing there already. phw == outsidecontext :D

RobertKrawitz commented:
To further update the comment: it appears that adding a tag to a missing track (with a pair of 8th notes next to it) is most likely to trigger a hang right now.
